pkg/dataconn/client.go

```diff
@@ -152,10 +115,17 @@ func (c *Client) Close() {
 }
 
 func (c *Client) loop() {
+	c.timer = time.NewTimer(opTimeout)
+	if !c.timer.Stop() {
+		<-c.timer.C
+	}
+
 	defer close(c.send)
 
 	for {
 		select {
+		case <-c.timer.C:
```

Comment:
In fact I am not sure when `select` will refresh the conditions if `c.timer` has changed.

Reply:
c.timer is only changed by one of the function called via the select.   select waits for one of the channels then it executes the function that can change c.timer.  c.timer isn't changed while the select is executing and waiting for the channels.   The channels that select is waiting for won't channel during the select.  The next iteration of select may be using different channels because c.timer was updated.